Tag API v1

i509VCB opened the pull request:
This PR deprecates the old V0 `tag-extensions` module and replaces it with `fabric-tag-api-v1`

~~This resolves #263~~
Extra tag types sadly impose a client and server mod requirement which is unacceptable for vanilla clients. I will open up my own libraries or future pull requests to handle that special case later.

Juuxel commented on fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/api/tag/extension/v1/FabricTag.java:
```diff
@@ -0,0 +1,30 @@
+ * Interface implemented by {@link net.minecraft.tag.Tag} instances when
+ * Fabric API is present.
+ *
+ * @param <T>
```
```suggestion
 * @param <T> the type contained in the tag
```

Juuxel commented on fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/api/tag/extension/v1/FabricTagBuilder.java:
```diff
@@ -0,0 +1,30 @@
+ * Interface implemented by {@link net.minecraft.tag.Tag.Builder} instances when
+ * Fabric API is present.
+ *
+ * @param <T>
```
```suggestion
 * @param <T> the type contained in the tag
```

Juuxel commented on fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/api/tag/extension/v1/TagRegistry.java:
```diff
@@ -0,0 +1,99 @@
+
+	private static Field fluidTagContainer;
+
+	private static TagContainer<Fluid> getFluidTagContainer() {
```
I think this can use a static `@Accessor`.

liach commented on fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/impl/tag/extension/v1/RegistryTagReloadListener.java:
```diff
@@ -0,0 +1,57 @@
+import net.fabricmc.fabric.api.resource.SimpleSynchronousResourceReloadListener;
+import net.fabricmc.fabric.api.tag.extension.v1.TagRegistry;
+
+public class RegistryTagReloadListener implements SimpleSynchronousResourceReloadListener {
```
Please make this async.

liach commented:
i think you made the api's scope too big. the old api is like plugin usable and now you add the networking, which really needs a client

i509VCB commented:
I could split the registry tags into a separate module if we feel the scope is now too big.

The reason I added the tag synchronization was to consider the fact that some mods on the client may want to sort enchantments by category or use biome tags.

EDIT: Obviously it's a draft just for that reason since the scope is still being decided.

liach commented:
after another look i think this setup is fine. just make sure you only send the packet if connection is not local and client can actually receive this packet type

i509VCB commented:
I need to rebase this

liach commented on build.gradle:
```diff
@@ -225,7 +235,7 @@ curseforge {
 		changelog = 'A changelog can be found at https://github.com/FabricMC/fabric/commits/master'
 		releaseType = 'beta'
-		addGameVersion '1.15-Snapshot'
+		addGameVersion '1.15.2'
```
Why? this appears wrong to me. Maybe you've messed up git

i509VCB replied:
something screwed up and I need to go look into why it messed up (probably cherrypick it to fix)

i509VCB replied:
Hmm due to the tag changes, should I count on doing this in 1.15 or 1.16.

I know modmuss had a blast backporting the object builders when I did them for 1.16 instead.

liach replied:
You probably shouldn't change the base branch of pr to 1.16; should change it back to 1.15, and cherry pick the squashed commit onto 1.16 and finally reset the head of `feature/registry-tags` to the repicked commit, then change the pr's base branch to 1.16

i509VCB commented:
I am reconsidering some design aspects of this PR. 1.16 having the changes with dimension types does make this a bit of fun.

i509VCB committed:
Start work on extra tags

(cherry picked from commit 295ed4732132d363b51b9526239e29d722a1a780)

Extra tag types had to go sadly. And 20w28a

Extra tag types are a fun issue in themselves. How do clients handle the fact a server may not have tags present. How does a client with required tags handle the server being vanilla? These questions sadly outweight the scope of this PR.

liach commented:
The tagging shouldn't hurt too much here. I expect dimension tags to work like function tags instead of other registry tags.

i509VCB commented:
I decided to bind dimension tags to the current dimension tracker, and the registry it holds. Locally I am working on it and I've decided so far to support the following:

```
Biomes
BlockEntityTypes
DimensionTypes
Enchantments
Painting(Motives/Motifs/Whatever)
ParticleTypes
PointOf InterestTypes
SoundEvents
StatusEffects
VillagerProfessions
VillagerTypes.
```

After some further realization, this will be waiting till modmuss finishes and commits his registry sync fixes, since I have to mark all those registries as `SYNCED` in order to have tags sync properly (since they use the raw ids in packets).

liach commented:
Why must you sync tags? Imo tag syncing should be optional (vanilla does not sync function tags for instance) and if a registry is int-id synced, we can sync tag elements by int id safely. No need to touch registry at all.

i509VCB commented:
I think the expectation is that these types of tags are synced. Function tags are the exception in this case since they are not bound to a registry.

On the case of optional or not, I only require the client to understand the packet in order to sync the custom tags. So a vanilla client would not have custom tags synced. ClientWorld does call some BlockTags checks, so I'd assume these tags should be synced to the client.

> if a registry is int-id synced, we can sync tag elements by int id safely. No need to touch registry at all.

To avoid syncing ids would require some custom packet logic to use identifiers instead. It's possible but I could see scenarios where we hit the maximum packet payload easily. (Even if we sent each type of tag container individually).

i509VCB commented:
Finally put my changes on branch. This is now targetting 1.16.

Note this is being cleaned up a bit before the final review cycle comes

i509VCB commented:
Okay so I did some experimenting with dimension types. The require changes are too invasive for fabric api.

This includes moving where RegistryTracker is initialized in `server.Main`, Replacing all calls to `ServerResourceManager.reload()` to a new call which does the same thing with a registry tracker.

So the current tagged types are final as they are.

Since Dimension tags are possible but not done easily right now, I will make a mod to add taggable dimensions and use that as a better gauge of whether it belongs in fabric api at a later date.

Will write up some final testing for this tomorrow and then I can open this up to review.

Prospector commented on fabric-extra-tag-api-v1/build.gradle:
```diff
@@ -0,0 +1,13 @@
+archivesBaseName = "fabric-registry-tag-api-v1"
```
Is it `registry-tag-api` or `extra-tag-api`? I'll be honest, I'm really not a fan of `extra-tag-api` but unfortunately I don't really have an alternative idea. Maybe `fabric-more-tag-types-api`?

i509VCB replied:
Tbf `registry-tag-api` was the original name, but the fact we don't cover every registry makes it misleading.

The name could be improved for sure.

Prospector commented on fabric-tag-api-v1/src/main/java/net/fabricmc/fabric/api/tag/v1/TagRegistry.java:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	public static Tag.Identified<Block> block(Identifier id) {
+		return getBlockAccessor().get(id.toString());
```
Why not just use an Invoker to invoke vanilla's `register` methods directly?

i509VCB replied:
both have the same cost anyways. From experience with registries, we should not use the default register methods due to defaulting to minecraft namespace.

i509VCB commented:
I need to update the module to use identified tags for initial release. Extra tag module at a later date probably.

i509VCB committed:
Cleanup from rebase

i509VCB commented:
Rebased to latest.

Juuxel reviewed (changes requested):
This looks really nice. Just one note: `tag-extensions-v0`'s version should be bumped.

(Also, couldn't you just use an `@Invoker` for the `BlockTags.register` method etc instead of accessing the `RequiredTagList`?)

i509VCB commented:
Are you referring to the `required` or regular method? These methods both do different things.

Juuxel commented:
The `required` method. At least `BlockTags.register` registers the tags as required tags (found this out when porting Adorn to use the vanilla system as I needed `Tag.Identified`).

i509VCB committed:
Use register method for required tags

i509VCB commented:
Moved to invoker on register, it is functionally the same as field access

i509VCB committed:
Some extra javadoc

i509VCB committed:
Internals yet again
